File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    void help(vector<int>& nums, int target, vector<vector<int>>& res, vector<int> subset, int idx){
        if(target==0){
            res.push_back(subset);
            return;
        }
        
        if(target<0 || idx>=nums.size()) return;
        
//         subset.push_back(nums[idx]);
//         help(nums, target-nums[idx], res, subset, idx);
//         subset.pop_back();
        
//         help(nums, target, res, subset, idx+1);
        
        
        for(int i=idx;i<nums.size();++i){
            subset.push_back(nums[i]);
            help(nums, target-nums[i],res, subset,i);
            subset.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> res{};
        vector<int> subset{};
        
        help(nums, target, res, subset, 0);
        return res;
    }
};

File: 39-combination-sum/39-combination-sum.java
class Solution {
    List<List<Integer>> res=new ArrayList<>();
    
    void solve(List<Integer> l, int[] nums, int t, int idx){
        if(idx==nums.length){
            if(t==0)
                res.add(new ArrayList<>(l));
            return;
        }
        
        if(t<0)
            return;
        
        l.add(nums[idx]);
        solve(l, nums, t-nums[idx], idx);
        l.remove(l.size()-1);
        
        solve(l, nums, t, idx+1);
    }
    
    public List<List<Integer>> combinationSum(int[] nums, int t) {
                
        solve(new ArrayList<Integer>(), nums, t, 0);
        return res;
    }
}